library/core/src/intrinsics.rs

```diff
@@ -2569,6 +2569,16 @@ extern "rust-intrinsic" {
     #[rustc_nounwind]
     pub fn is_val_statically_known<T: Copy>(arg: T) -> bool;
 
+    /// Returns the value of `cfg!(debug_assertions)`, but after monomorphization instead of in
+    /// macro expansion.
+    ///
+    /// This always returns `false` in const eval (including Miri). The interpreter provides better
+    /// diagnostics than the checks that this is used to implement.
```

Comment:
```suggestion
    /// diagnostics than the checks that this is used to implement. However, this means
    /// you should only be using this intrinsic to guard requirements that, if violated,
    /// immediately lead to UB. Otherwise, const-eval and Miri will miss out on those
    /// checks entirely.
```